Lib/subprocess.py

```diff
@@ -1245,11 +1245,39 @@ def _check_timeout(self, endtime, orig_timeout, stdout_seq, stderr_seq,
         """Convenience for checking if a timeout has expired."""
         if endtime is None:
             return
+
+        def translate_newlines_partial_output(data, encoding, errors):
+            # Handle decoding the data considering it may be truncated
+            # mid-codepoint (ignore the trailing partial codepoint).
+            # See https://github.com/python/cpython/issues/87597
+            try:
+                output = self._translate_newlines(data, encoding, errors)
+            except UnicodeDecodeError as exc:
+                if exc.end == len(data):
+                    output = self._translate_newlines(data[:exc.start],
+                                                      encoding,
+                                                      errors)
+                else:
+                    raise
```

Comment:
This exception would be raised instead of the timeout error, which isn't going to be helpful. I considered three alternatives, but I think the last is the best:

* start raising the TimeoutExpired, then raise the UnicodeDecodeError `from` it
* defer decoding to the `TimeoutExpired` class (so it learns to keep encoding/errors around and do the decode on demand)
* decode up to `exc.start` here, then decode the rest with the `replace` error handler and concatenate it

Reply:
Agreed, I like that latter strategy. I think it is important to have the `TimeoutExpired` be the primary error. Otherwise the caller could wrongly interpret the `UnicodeDecodeError` as the process exiting naturally while producing undecodable output.  If the process did emit undecodable data on its own, that isn't important in this situation.